### PR TITLE
Add CORS preflight handler, enlarge upload file size and add docker-compose.yaml

### PR DESCRIPTION
**Pull request by FilJed**
First of all -- thank you for this usefull reverse proxy. I have made a few modifications to your project. Please review my code and let me know if you have any feedback or suggestions. Thank you!

# Problems addressed
## CORS
API gateway functions properly, but when using it in e.g. browser pluggins (or obsidian) it breaks on CORS preflight checks:
```
Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

## Nginx file size default limitations 
By default, Nginx has a file size limit of 1mb, whereas the OpenAI API allows file sizes up to 25mb. This becomes problematic when using the Whisper API for tasks such as audio transcription, as the current limit is quickly reached.

# Changes added
- add CORS handlers 
- expande file size limit to 25mb
- include docker-compose.yaml file

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+version: '3'
+services:
+  openai-api-gateway:
+    build: .
+    container_name: openai-api-gateway
+    ports:
+      - 80:40000
```
Comment (BadgerHobbs) [inline docker-compose.yaml:7]: This should be port 80

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    environment:
+      - GATEWAY_API_KEY=super_secret_api_key                               # Required
+      - OPENAI_API_KEY=$OPENAI_API_KEY                                     # Required
+      - OPENAI_API_HOST=https://api.openai.com                             # Optional
+      - RATE_LIMIT=20r/m                                                   # Optional
+    restart: unless-stopped
```
Comment (BadgerHobbs) [inline docker-compose.yaml:13]: File just needs a new line at the end (GitHub complains)

**Comment by BadgerHobbs** [comment]
Nice job, thanks for the contribution and raising these issues, I've left a couple quick comments and shall do a full test/review this evening.

File: gateway-nginx.conf [hunk 3]
```diff
@@ -2,20 +2,35 @@
 # Define a shared memory zone "api" to store states of limiting requests. The maximum size for this zone is 10MB. RATE_LIMIT sets the rate of requests
 limit_req_zone $binary_remote_addr zone=api:10m rate=${RATE_LIMIT};
 
+# OpenAi API limitation is 25mb
+client_max_body_size 25m;
```
Comment (BadgerHobbs) [inline gateway-nginx.conf:6]: For future compatibility/flexibility, could replace the value with an environment variable like for the rate limit and such.

```
client_max_body_size ${CLIENT_MAX_BODY_SIZE};
```

Dockerfile
```dockerfile
ENV CLIENT_MAX_BODY_SIZE=25m
```

**Review by BadgerHobbs — CHANGES_REQUESTED** [review]

File: gateway-nginx.conf (hunk 3 above)
Comment (BadgerHobbs) [reply gateway-nginx.conf:6]: I've gone ahead and updated your implementation to have all the CORS settings ad environment variables with sensible defaults, see below.

gateway-nginx.conf
```sh

# Define a shared memory zone "api" to store states of limiting requests. The maximum size for this zone is 10MB. RATE_LIMIT sets the rate of requests
limit_req_zone $binary_remote_addr zone=api:10m rate=${RATE_LIMIT};

# Set response body size (OpenAI API supports 25MB)
client_max_body_size ${CLIENT_MAX_BODY_SIZE};

# Start server configuration
server {

    # Configure Nginx to listen on port 80
    listen 80;

    # "location /" defines how to respond to requests for URLs beginning with / 
    location / {

        # Pre-flight requests (OPTIONS)
        if ($request_method = "OPTIONS") {
            add_header "Access-Control-Allow-Origin" "${CORS_ORIGIN}";
            add_header "Access-Control-Allow-Methods" "${CORS_METHODS}";
            add_header "Access-Control-Allow-Headers" "${CORS_HEADERS}";
            add_header "Access-Control-Allow-Credentials" "${CORS_CREDENTIALS}";
            add_header "Access-Control-Max-Age" ${CORS_MAX_AGE};
            add_header "Content-Length" ${CONTENT_LENGTH};
            add_header "Content-Type" "${CONTENT_TYPE}";
            return 204;
        }

        # Respond with CORS headers for all other requests
        add_header "Access-Control-Allow-Origin" "${CORS_ORIGIN}";
        add_header "Access-Control-Allow-Methods" "${CORS_METHODS}";
        add_header "Access-Control-Allow-Headers" "${CORS_HEADERS}";
        add_header "Access-Control-Allow-Credentials" "${CORS_CREDENTIALS}";

        # If the client is not authorized ($http_authorization not equals to "Bearer ${GATEWAY_API_KEY}"), return an HTTP 401 Unauthorized status code
        if ($http_authorization != "Bearer ${GATEWAY_API_KEY}") {
            return 401;
        }
        
        # Set the proxy server to be used for requests
        proxy_pass ${OPENAI_API_HOST};

        # Set the header fields for requests being proxied
        proxy_set_header Authorization "Bearer ${OPENAI_API_KEY}";
        proxy_set_header api-key "${OPENAI_API_KEY}";

        # Turn off buffering of responses from the proxy server
        proxy_buffering off;

        # Transfer the host name from the request line to the SSL handshake
        proxy_ssl_server_name on;

        # Enable limiting of requests by defining its parameters with this directive
        limit_req zone=api;

        # Return client status 429 (Too Many Requests) when the rate of requests exceeds the defined limit
        limit_req_status 429;
    }
}
```

Dockerfile
```dockerfile
FROM nginx:latest

# Default environment variable values
ENV OPENAI_API_HOST="https://api.openai.com"
ENV RATE_LIMIT="20r/m"
ENV CLIENT_MAX_BODY_SIZE="25m"
ENV CORS_ORIGIN="*"
ENV CORS_METHODS="*"
ENV CORS_HEADERS="*"
ENV CORS_CREDENTIALS="true"
ENV CORS_MAX_AGE="86400"
ENV CONTENT_LENGTH="0"
ENV CONTENT_TYPE="text/plain charset=UTF-8"

# Copy gateway configuration
COPY gateway-nginx.conf /etc/nginx/conf.d_template/default.conf

# Copy and run init script to inject environment variables into configuration and run nginx
COPY init.sh /docker-entrypoint.d/init.sh
RUN chmod +x /docker-entrypoint.d/init.sh
```

init.sh
```sh
#!/bin/sh

# Substitute environment variables
envsubst '${GATEWAY_API_KEY},${OPENAI_API_KEY},${OPENAI_API_HOST},${AZURE_OPENAI_API},${RATE_LIMIT},${CLIENT_MAX_BODY_SIZE},${CORS_ORIGIN},${CORS_METHODS},${CORS_HEADERS},${CORS_CREDENTIALS},${CORS_MAX_AGE},${CONTENT_LENGTH},${CONTENT_TYPE}' < /etc/nginx/conf.d_template/default.conf > /etc/nginx/conf.d/default.conf

# Start nginx
exec "nginx" "-g" "daemon off;"
```

readme
##### OpenAI API
```sh
docker run -d \
    --name openai-api-gateway \
    -p 80:80 \
    -e GATEWAY_API_KEY=super_secret_api_key \       # Required
    -e OPENAI_API_KEY=your_api_key_here \           # Required
    -e OPENAI_API_HOST=https://api.openai.com \     # Optional
    -e RATE_LIMIT=20r/m \                           # Optional
    -e CLIENT_MAX_BODY_SIZE=25m \                   # Optional
    -e CORS_ORIGIN=* \                              # Optional
    -e CORS_METHODS=* \                             # Optional
    -e CORS_HEADERS=* \                             # Optional
    -e CORS_CREDENTIALS=true \                      # Optional
    -e CORS_MAX_AGE=86400 \                         # Optional
    -e CONTENT_LENGTH=0 \                           # Optional
    -e CONTENT_TYPE=text/plain charset=UTF-8 \      # Optional
    ghcr.io/badgerhobbs/openai-api-gateway:main
```

##### Azure OpenAI API
```sh
docker run -d \
    --name azure-openai-api-gateway \
    -p 80:80 \
    -e GATEWAY_API_KEY=super_secret_api_key \                               # Required
    -e OPENAI_API_KEY=your_api_key_here \                                   # Required
    -e OPENAI_API_HOST=https://my-azure-openai-api.openai.azure.com \       # Required
    -e RATE_LIMIT=20r/m \                                                   # Optional
    -e CLIENT_MAX_BODY_SIZE=25m \                                           # Optional
    -e CORS_ORIGIN=* \                                                      # Optional
    -e CORS_METHODS=* \                                                     # Optional
    -e CORS_HEADERS=* \                                                     # Optional
    -e CORS_CREDENTIALS=true \                                              # Optional
    -e CORS_MAX_AGE=86400 \                                                 # Optional
    -e CONTENT_LENGTH=0 \                                                   # Optional
    -e CONTENT_TYPE=text/plain charset=UTF-8 \                              # Optional
    ghcr.io/badgerhobbs/openai-api-gateway:main
```

#### Docker Compose

The following is an example docker compose file for deploying gateways for both the OpenAI and Azure OpenAI API.

```yaml
version: '3'
services:

  openai-api-gateway:
    image: ghcr.io/badgerhobbs/openai-api-gateway:main
    container_name: openai-api-gateway
    ports:
      - 80:80
    environment:
      - GATEWAY_API_KEY=super_secret_api_key                              # Required
      - OPENAI_API_KEY=your_api_key_here                                  # Required
      - OPENAI_API_HOST=https://api.openai.com                            # Optional
      - RATE_LIMIT=20r/m                                                  # Optional
      - CLIENT_MAX_BODY_SIZE=25m                                          # Optional
      - CORS_ORIGIN=*                                                     # Optional
      - CORS_METHODS=*                                                    # Optional
      - CORS_HEADERS=*                                                    # Optional
      - CORS_CREDENTIALS=true                                             # Optional
      - CORS_MAX_AGE=86400                                                # Optional
      - CONTENT_LENGTH=0                                                  # Optional
      - CONTENT_TYPE=text/plain charset=UTF-8                             # Optional
    restart: unless-stopped

  azure-openai-api-gateway:
    image: ghcr.io/badgerhobbs/openai-api-gateway:main
    container_name: azure-openai-api-gateway
    ports:
      - 80:80
    environment:
      - GATEWAY_API_KEY=super_secret_api_key                              # Required
      - OPENAI_API_KEY=your_api_key_here                                  # Required
      - OPENAI_API_HOST=https://my-azure-openai-api.openai.azure.com      # Required
      - RATE_LIMIT=20r/m                                                  # Optional
      - CLIENT_MAX_BODY_SIZE=25m                                          # Optional
      - CORS_ORIGIN=*                                                     # Optional
      - CORS_METHODS=*                                                    # Optional
      - CORS_HEADERS=*                                                    # Optional
      - CORS_CREDENTIALS=true                                             # Optional
      - CORS_MAX_AGE=86400                                                # Optional
      - CONTENT_LENGTH=0                                                  # Optional
      - CONTENT_TYPE=text/plain charset=UTF-8                             # Optional
    restart: unless-stopped
```

**Comment by BadgerHobbs** [comment]
Thanks again for your contribution on this.

I have created another PR #3 branching off your changes and appending my comments. As such closing this one.

**Comment by FilJed** [comment]
Wow, the work you done, it's amazing!
Sorry for delayed response -- been out of service.
But I see, there's no contribution left to make. 
Cheers!

**Pull request closed without merging**